File: Day12/day12.py
import sys

ORD_MIN = ord("a")
ORD_MAX = ord("z")
ALTITUDE_DICT = {"S": 0, "E": ORD_MAX - ORD_MIN}
for element in "abcdefghijklmnopqrstuvwxyz":
    ALTITUDE_DICT[element] = ord(element) - ORD_MIN


def row_col_to_pos(row: int, column: int, number_rows: int, number_columns: int) -> int:
    return row * number_columns + column


def pos_to_col_and_row(position: int, number_rows: int, number_columns: int) -> int:
    row = position // number_columns
    column = position % number_columns
    return (row, column)


def check_neighbours(
    elevation_map: str, position: int, number_rows: int, number_cols: int
) -> list[list[int]]:
    neighbours = []
    current_altitude = ALTITUDE_DICT[elevation_map[position]]
    number_elements = number_rows * number_cols
    row, column = pos_to_col_and_row(position, number_rows, number_cols)

    top = row_col_to_pos(row - 1, column, number_rows, number_cols)
    if top >= 0 and ALTITUDE_DICT[elevation_map[top]] >= current_altitude - 1:
        neighbours.append(top)

    left = position - 1
    if left >= 0 and ALTITUDE_DICT[elevation_map[left]] >= current_altitude - 1:
        neighbours.append(left)

    bottom = row_col_to_pos(row + 1, column, number_rows, number_cols)
    if (
        bottom < number_elements
        and ALTITUDE_DICT[elevation_map[bottom]] >= current_altitude - 1
    ):
        neighbours.append(bottom)

    right = position + 1
    if (
        right < number_elements
        and ALTITUDE_DICT[elevation_map[right]] >= current_altitude - 1
    ):
        neighbours.append(right)

    return neighbours


class Terrain:
    def __init__(self, elevation_map: str):
        global ALTITUDE_DICT
        split_map = elevation_map.split("\n")
        split_map.remove("")
        self.distances_computed = False
        self.number_rows = len(split_map)
        self.number_cols = len(split_map[0])
        elevation_map = elevation_map.replace("\n", "")
        self.positions = []
        number_elements = len(elevation_map)

        for index, element in enumerate(elevation_map):
            self.positions.append(
                Position(ALTITUDE_DICT[element], index, number_elements)
            )
            if element == "S":
                self.start = index
            if element == "E":
                self.end = index
                self.positions[-1].distance_to_end = 0
            self.positions[-1].neighbours = check_neighbours(
                elevation_map, index, self.number_rows, self.number_cols
            )

    def compute_distances(self):
        self.distances_computed = True
        unvisited = list(range(len(self.positions)))
        current_position = self.end
        max_distance = self.positions[self.start].distance_to_end
        while unvisited != []:
            if current_position == self.start:
                break
            tentative_distance = self.positions[current_position].distance_to_end + 1
            for neighbour in [
                neighbour
                for neighbour in self.positions[current_position].neighbours
                if neighbour in unvisited
            ]:
                if tentative_distance < self.positions[neighbour].distance_to_end:
                    self.positions[neighbour].distance_to_end = tentative_distance
            unvisited.remove(current_position)
            min_unvisited_distance = max_distance
            for position in unvisited:
                if self.positions[position].distance_to_end <= min_unvisited_distance:
                    current_position = position
                    min_unvisited_distance = self.positions[position].distance_to_end

    def distance(self):
        if not self.distances_computed:
            self.compute_distances()
        return self.positions[self.start].distance_to_end

    def min_distance_from_bottom(self) -> int:
        if not self.distances_computed:
            self.compute_distances()
        distance = self.number_rows * self.number_cols
        for position in self.positions:
            if position.altitude == 0:
                distance = min(distance, position.distance_to_end)
        return distance


class Position:
    def __init__(self, altitude: int, position: int, distance_to_end: int):
        self.altitude = altitude
        self.neighbours = []
        self.position = position
        self.distance_to_end = distance_to_end

    def __repr__(self):
        return f"Position at {self.position} of altitude {self.altitude}."


if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
    except IndexError:
        file_name = "input.txt"
    elevation_map = open(file_name).read()
    terrain_map = Terrain(elevation_map)

    print(f"The shortest path has length {terrain_map.distance()}.")
    print(
        f"The closest point from the botton is at {terrain_map.min_distance_from_bottom()} steps from the end."
    )

File: Day12/day12.rs
const VALUE_A: u32 = b'a' as u32;

fn get_data() -> String {
    let input_args: Vec<String> = std::env::args().collect();
    let mut file_name: String = "input.txt".to_string();
    if input_args.len() > 1 {
        file_name = input_args[1].clone();
    }
    let mut data = std::fs::read_to_string(file_name).unwrap();
    if data.ends_with("\n") {
        data.pop();
    }
    return data;
}

fn row_col_to_pos(row: &usize, col: &usize, number_cols: &usize) -> usize {
    row * number_cols + col
}

fn pos_to_row_col(pos: &usize, number_cols: &usize) -> (usize, usize) {
    (pos / number_cols, pos.rem_euclid(*number_cols))
}

fn check_neiboughrs(
    elevation_map: &[u8],
    position: &usize,
    number_rows: &usize,
    number_cols: &usize,
) -> Vec<usize> {
    let mut neighbours = vec![];
    let current_altitude = elevation_map[*position] as u32 - VALUE_A;
    let number_elements = number_rows * number_cols;
    let (row, column) = pos_to_row_col(&position, &number_cols);

    if row > 0 {
        let top = row_col_to_pos(&(row - 1), &column, &number_cols);
        if (elevation_map[top] as u32) >= current_altitude + VALUE_A - 1 {
            neighbours.push(top);
        }
    }

    if position > &0 {
        let left = position - 1;
        if (elevation_map[left] as u32) >= current_altitude + VALUE_A - 1 {
            neighbours.push(left);
        }
    }

    let bottom = row_col_to_pos(&(row + 1), &column, &number_cols);
    if bottom < number_elements {
        if (elevation_map[bottom] as u32) >= current_altitude + VALUE_A - 1 {
        neighbours.push(bottom);
        }
    }

    let right = position + 1;
    if right < number_elements {
        if (elevation_map[right] as u32) >= current_altitude + VALUE_A - 1 {
            neighbours.push(right);
        }
    }

    neighbours
}

struct Position {
    altitude: u32,
    neighbours: Vec<usize>,
    _position: usize,
    distance_to_end: u32,
}

struct Terrain {
    number_rows: usize,
    number_cols: usize,
    positions: Vec<Position>,
    start: usize,
    end: usize,
    distances_computed: bool,
}

impl Terrain {
    fn build_terrain(&mut self, elevation_map: String) {
        let split_map: Vec<&str> = elevation_map.split("\n").collect();
        self.number_rows = split_map.len();
        self.number_cols = split_map[0].len();
        let elevation_map_stripped = str::replace(&elevation_map, "\n", "");
        let elevation_only_map = elevation_map_stripped.replace("S", "a").replace("E", "z");
        let elevation_map_as_bytes = elevation_only_map.as_bytes();
        let number_elements = (self.number_cols * self.number_rows) as u32;
        let mut distance_to_end = 0;
        let mut altitude;

        for (index, element) in elevation_map_stripped.as_bytes().iter().enumerate() {
            if element == &b'S' {
                self.start = index;
                altitude = b'a' as u32 - VALUE_A;
            } else if element == &b'E' {
                self.end = index;
                distance_to_end = 0;
                altitude = b'z' as u32 - VALUE_A;
            } else {
                distance_to_end = number_elements;
                altitude = *element as u32 - VALUE_A;
            }

            let neighbours = check_neiboughrs(
                &elevation_map_as_bytes,
                &index,
                &self.number_rows,
                &self.number_cols,
            );

            self.positions.push(Position {
                altitude: altitude,
                neighbours: neighbours,
                _position: index,
                distance_to_end: distance_to_end,
            });
        }
    }

    fn distance_start_to_end(&mut self) -> u32 {
        if self.distances_computed == false {
            self.compute_distances();
        }
        self.positions[self.start].distance_to_end
    }

    fn min_distance_from_bottom(&mut self) -> u32 {
        if self.distances_computed == false {
            self.compute_distances()
        }
        let mut distance = (self.number_rows * self.number_cols) as u32;
        for position in &self.positions {
            if position.altitude == 0 {
                distance = std::cmp::min(distance, position.distance_to_end)
            }
        }
        distance
    }

    fn compute_distances(&mut self) {
        self.distances_computed = true;
        let mut unvisited: Vec<usize> = (0..self.positions.len()).collect();
        let mut current_position = self.end;
        let mut tentative_distance;
        let max_distance = self.positions[self.start].distance_to_end;
        let mut min_unvisited_distance;
        while unvisited.len() > 0 {
            if current_position == self.start {
                break;
            }
            tentative_distance = self.positions[current_position].distance_to_end + 1;
            let current_neighbours = self.positions[current_position].neighbours.clone();
            for neighbour in current_neighbours {
                if unvisited.contains(&neighbour)
                    & (tentative_distance < self.positions[neighbour].distance_to_end)
                {
                    self.positions[neighbour].distance_to_end = tentative_distance;
                }
            }
            unvisited.remove(unvisited.iter().position(|x| *x == current_position).expect("Element not found"));
            min_unvisited_distance = max_distance;
            for position in &unvisited {
                if self.positions[*position].distance_to_end <= min_unvisited_distance {
                    current_position = *position;
                    min_unvisited_distance = self.positions[*position].distance_to_end;
                }
            }
        }
    }
}

fn main() {
    let data = get_data();
    let mut terrain_map = Terrain{
        number_rows: 0,
        number_cols: 0,
        positions: vec![],
        start: 0,
        end: 0,
        distances_computed: false,        
    };
    terrain_map.build_terrain(data);

    println!("The shortest path has length {}.", terrain_map.distance_start_to_end());
    println!("The closest point from the bottom is {} steps from the end.", terrain_map.min_distance_from_bottom());
}
